src/AspNet.Security.OAuth.WeChat/WeChatAuthenticationHandler.cs

```diff
@@ -0,0 +1,147 @@
+{
+    public class WeChatAuthenticationHandler : OAuthHandler<WeChatAuthenticationOptions>
+    {
+        public WeChatAuthenticationHandler(
+            [NotNull] IOptionsMonitor<WeChatAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity, [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var payload = new JObject();
+            switch (Options.Scope.SingleOrDefault())
+            {
+                case "snsapi_base":
+                    payload["openid"] = tokens.Response.Value<string>("openid");
+                    break;
+                case "snsapi_userinfo":
+                    var address = QueryHelpers.AddQueryString(Options.UserInformationEndpoint,
+                        new Dictionary<string, string>
+                        {
+                            ["access_token"] = tokens.AccessToken,
+                            ["openid"] = tokens.Response.Value<string>("openid")
+                        });
+
+                    var response = await Backchannel.GetAsync(address);
+                    if (!response.IsSuccessStatusCode)
+                    {
+                        Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                        "returned a {Status} response with the following payload: {Headers} {Body}.",
+                            /* Status: */ response.StatusCode,
+                            /* Headers: */ response.Headers.ToString(),
+                            /* Body: */ await response.Content.ReadAsStringAsync());
+
+                        throw new HttpRequestException("An error occurred while retrieving user information.");
+                    }
+
+                    payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+                    if (!string.IsNullOrEmpty(payload.Value<string>("errcode")))
+                    {
+                        Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                        "returned a {Status} response with the following payload: {Headers} {Body}.",
+                            /* Status: */ response.StatusCode,
+                            /* Headers: */ response.Headers.ToString(),
+                            /* Body: */ await response.Content.ReadAsStringAsync());
+
+                        throw new HttpRequestException("An error occurred while retrieving user information.");
+                    }
+
+                    break;
+                default:
+                    throw new InvalidOperationException("Invalid scope");
```

Comment:
It would be useful to include what the invalid scope was in the exception message.